AsyncDisplayKit/Details/ASTableLayoutController.m

```diff
@@ -0,0 +1,147 @@
+@end
+
+@implementation ASTableLayoutController
+
+- (instancetype)initWithTableView:(UITableView *)tableView
+{
+  if (!(self = [super init])) {
+    return nil;
+  }
+  _tableView = tableView;
+  return self;
+}
+
+#pragma mark - Visible Indices
+
+- (void)setVisibleNodeIndexPaths:(NSArray *)indexPaths
+{
+  _visibleIndexPaths = [indexPaths sortedArrayUsingSelector:@selector(compare:)];
+}
+
+/**
+ * IndexPath array for the element in the working range.
+ */
+
+- (NSSet *)indexPathsForScrolling:(ASScrollDirection)scrollDirection rangeMode:(ASLayoutRangeMode)rangeMode rangeType:(ASLayoutRangeType)rangeType
+{
+  if (_visibleIndexPaths.count == 0) {
+    return [NSSet set];
+  }
+
+  CGSize viewportSize = [self viewportSize];
+
+  ASRangeTuningParameters tuningParameters = [self tuningParametersForRangeMode:rangeMode rangeType:rangeType];
+
+  ASDirectionalScreenfulBuffer directionalBuffer = ASDirectionalScreenfulBufferVertical(scrollDirection, tuningParameters);
+  
+  NSIndexPath *startPath = [self findIndexPathAtDistance:(-directionalBuffer.negativeDirection * viewportSize.height)
+                                          fromIndexPath:_visibleIndexPaths.firstObject];
+  
+  NSIndexPath *endPath   = [self findIndexPathAtDistance:(directionalBuffer.positiveDirection * viewportSize.height)
+                                          fromIndexPath:_visibleIndexPaths.lastObject];
+
+  NSSet *indexPaths = [self indexPathsFromIndexPath:startPath toIndexPath:endPath];
+  return indexPaths;
+}
+
+#pragma mark - Utility
+
+- (NSIndexPath *)findIndexPathAtDistance:(CGFloat)distance fromIndexPath:(NSIndexPath *)start
+{
+  CGFloat targetY = distance + CGRectGetMidY([_tableView rectForRowAtIndexPath:start]);
+
+  // Before first row.
+  NSIndexPath *firstIndexPath = [self firstIndexPathInTableView];
+  if (targetY <= CGRectGetMaxY([_tableView rectForRowAtIndexPath:firstIndexPath])) {
+    return firstIndexPath;
+  }
+
+  // After last row.
+  NSIndexPath *lastIndexPath = [self lastIndexPathInTableView];
+  if (targetY >= CGRectGetMinY([_tableView rectForRowAtIndexPath:lastIndexPath])) {
+    return lastIndexPath;
+  }
+
+  static CGFloat kInitialRowSearchHeight = 100;
+  /**
+   * There may not be a row at any given EXACT point, for these possible reasons:
+   * - There is a section header/footer at that point.
+   * - That point is beyond the start/end of the table content. (Handled above)
+   *
+   * Solution: Make a target rect, starting at height 100, and if we don't
+   * find any rows, keep doubling its height and searching again. In practice,
+   * this will virtually always find a row on the first try (unless you have a 
+   * section header more than 100 tall and we land near the middle.)
+   */
+  NSIndexPath *result = nil;
+  for (CGRect targetRect = CGRectMake(0, targetY, 1, kInitialRowSearchHeight);
+       result == nil;
+       targetRect = CGRectInset(targetRect, 0, -CGRectGetHeight(targetRect) / 2.0)) {
```

Comment:
Good point – we should probably err on the side of _more_ than ~2 screenfuls rather than less, so I've updated it to always search at the furthest extent in the aiming direction.